File: project-analyzer-java/src/main/java/com/analyzer/module/BusinessAnalyzer.java
package com.analyzer.module;

import com.analyzer.model.*;
import com.analyzer.model.AnalysisReport.*;
import com.analyzer.plugin.AnalysisModuleInterface;
import com.analyzer.plugin.LanguagePlugin;

import java.nio.file.Path;
import java.util.*;
import java.util.regex.Pattern;

public class BusinessAnalyzer implements AnalysisModuleInterface {

    private record ModelPattern(Pattern pattern, DataModelInfo.DataModelType type) {}

    private static final List<ModelPattern> DATA_MODEL_PATTERNS = List.of(
            new ModelPattern(Pattern.compile("entity", Pattern.CASE_INSENSITIVE), DataModelInfo.DataModelType.ENTITY),
            new ModelPattern(Pattern.compile("model", Pattern.CASE_INSENSITIVE), DataModelInfo.DataModelType.MODEL),
            new ModelPattern(Pattern.compile("dto", Pattern.CASE_INSENSITIVE), DataModelInfo.DataModelType.DTO),
            new ModelPattern(Pattern.compile("vo$", Pattern.CASE_INSENSITIVE), DataModelInfo.DataModelType.VO),
            new ModelPattern(Pattern.compile("VO[A-Z]", Pattern.CASE_INSENSITIVE), DataModelInfo.DataModelType.VO)
    );

    private static final Set<String> NOISE_WORDS = Set.of("src", "lib", "app", "main", "java", "kotlin");

    @Override
    public String getName() { return "business"; }

    @Override
    public Object analyze(ProjectProfile profile, List<LanguagePlugin> plugins) {
        List<ModuleInfo> modules = collectModules(profile, plugins);
        enrichModulesWithClasses(modules, plugins);
        inferDescriptions(modules);
        List<DataModelInfo> dataModels = extractDataModels(profile, plugins);

        BusinessResult result = new BusinessResult();
        result.setModules(modules);
        result.setDataModels(dataModels);
        return result;
    }

    private List<ModuleInfo> collectModules(ProjectProfile profile, List<LanguagePlugin> plugins) {
        List<ModuleInfo> modules = new ArrayList<>();
        Set<String> seen = new HashSet<>();
        for (LanguagePlugin plugin : plugins) {
            for (ModuleInfo mod : plugin.identifyModules(Path.of(profile.getProjectPath()))) {
                String normalized = Path.of(mod.getPath()).toAbsolutePath().normalize().toString();
                if (seen.add(normalized)) {
                    mod.setInferred(true);
                    modules.add(mod);
                }
            }
        }
        return modules;
    }

    private void enrichModulesWithClasses(List<ModuleInfo> modules, List<LanguagePlugin> plugins) {
        for (ModuleInfo mod : modules) {
            Set<String> classNames = new LinkedHashSet<>();
            for (String filePath : mod.getKeyFiles()) {
                for (LanguagePlugin plugin : plugins) {
                    for (AstNode node : plugin.parseFile(Path.of(filePath))) {
                        if (node.getType() == AstNode.AstNodeType.CLASS
                                || node.getType() == AstNode.AstNodeType.INTERFACE) {
                            classNames.add(node.getName());
                        }
                    }
                }
            }
            if (!classNames.isEmpty()) {
                mod.setKeyClasses(new ArrayList<>(classNames));
            }
        }
    }

    private void inferDescriptions(List<ModuleInfo> modules) {
        for (ModuleInfo mod : modules) {
            if (mod.getDescription() == null || mod.getDescription().startsWith("Module inferred")
                    || mod.getDescription().startsWith("基于目录推断的模块")
                    || mod.getDescription().startsWith("Module:") || mod.getDescription().startsWith("模块:")
                    || mod.getDescription().startsWith("Java package:") || mod.getDescription().startsWith("Java 包:")) {
                mod.setDescription(generateDescription(mod));
                mod.setInferred(true);
            }
        }
    }

    private String generateDescription(ModuleInfo mod) {
        String humanName = humanize(mod.getName());
        if (mod.getKeyClasses() != null && !mod.getKeyClasses().isEmpty()) {
            String hints = String.join(", ", mod.getKeyClasses().subList(0,
                    Math.min(3, mod.getKeyClasses().size())));
            return humanName + " 模块（核心类: " + hints + "）";
        }
        return humanName + " 模块（基于目录结构推断）";
    }

    private String humanize(String name) {
        String[] parts = name.replace("-", " ").replace("_", " ").replace(".", " ").split("\\s+");
        StringBuilder sb = new StringBuilder();
        for (String w : parts) {
            if (NOISE_WORDS.contains(w.toLowerCase())) continue;
            if (!sb.isEmpty()) sb.append(" ");
            sb.append(Character.toUpperCase(w.charAt(0))).append(w.substring(1));
        }
        return sb.isEmpty() ? name : sb.toString();
    }

    private List<DataModelInfo> extractDataModels(ProjectProfile profile, List<LanguagePlugin> plugins) {
        List<DataModelInfo> models = new ArrayList<>();
        Set<String> seen = new HashSet<>();
        Set<String> allFiles = new LinkedHashSet<>();
        for (LanguagePlugin plugin : plugins) {
            for (ModuleInfo mod : plugin.identifyModules(Path.of(profile.getProjectPath()))) {
                allFiles.addAll(mod.getKeyFiles());
            }
        }
        for (String filePath : allFiles) {
            for (LanguagePlugin plugin : plugins) {
                for (AstNode node : plugin.parseFile(Path.of(filePath))) {
                    if (node.getType() != AstNode.AstNodeType.CLASS
                            && node.getType() != AstNode.AstNodeType.INTERFACE) continue;
                    DataModelInfo.DataModelType modelType = classifyModelType(node.getName(), filePath);
                    if (modelType == null) continue;
                    String key = filePath + ":" + node.getName();
                    if (!seen.add(key)) continue;

                    List<DataFieldInfo> fields = extractFields(node);
                    DataModelInfo dmi = new DataModelInfo();
                    dmi.setName(node.getName());
                    dmi.setType(modelType);
                    dmi.setFilePath(filePath);
                    dmi.setFields(fields);
                    dmi.setDescription(modelType.getValue().toUpperCase() + " 类: " + node.getName());
                    models.add(dmi);
                }
            }
        }
        return models;
    }

    private DataModelInfo.DataModelType classifyModelType(String className, String filePath) {
        for (ModelPattern mp : DATA_MODEL_PATTERNS) {
            if (mp.pattern().matcher(className).find()) return mp.type();
        }
        String baseName = Path.of(filePath).getFileName().toString();
        int dot = baseName.lastIndexOf('.');
        if (dot > 0) baseName = baseName.substring(0, dot);
        for (ModelPattern mp : DATA_MODEL_PATTERNS) {
            if (mp.pattern().matcher(baseName).find()) return mp.type();
        }
        return null;
    }

    private List<DataFieldInfo> extractFields(AstNode node) {
        List<DataFieldInfo> fields = new ArrayList<>();
        if (node.getChildren() == null) return fields;
        for (AstNode child : node.getChildren()) {
            if (child.getType() != AstNode.AstNodeType.FIELD) continue;
            DataFieldInfo fi = new DataFieldInfo();
            fi.setName(child.getName());
            fi.setType(child.getReturnType() != null ? child.getReturnType() : "unknown");
            fi.setAnnotations(child.getAnnotations() != null
                    ? child.getAnnotations().stream().map(AstNode.Annotation::getName).toList()
                    : List.of());
            fields.add(fi);
        }
        return fields;
    }
}

File: project-analyzer-ts/src/modules/business-analyzer.ts
/**
 * BusinessAnalyzer — 业务功能分析模块
 *
 * 基于目录结构识别业务功能模块，提取关键类和文件列表，
 * 推断模块功能描述，并提取数据模型（Entity/Model/DTO/VO）。
 */

import * as path from 'node:path';
import type { AnalysisModuleInterface, ModuleResult } from '../plugins/analysis-module.js';
import type { LanguagePlugin } from '../plugins/language-plugin.js';
import type {
  ProjectProfile,
  ModuleInfo,
  BusinessResult,
  DataModelInfo,
  DataFieldInfo,
  AstNode,
} from '../models/index.js';

/** Patterns that indicate a data model class. */
const DATA_MODEL_PATTERNS: { pattern: RegExp; type: DataModelInfo['type'] }[] = [
  { pattern: /entity/i, type: 'entity' },
  { pattern: /model/i, type: 'model' },
  { pattern: /dto/i, type: 'dto' },
  { pattern: /vo$/i, type: 'vo' },
  { pattern: /VO[A-Z]/i, type: 'vo' },
];

/** Common words to strip from directory names when generating descriptions. */
const NOISE_WORDS = new Set(['src', 'lib', 'app', 'main', 'java', 'kotlin']);

export class BusinessAnalyzer implements AnalysisModuleInterface {
  getName(): string {
    return 'business';
  }

  async analyze(profile: ProjectProfile, plugins: LanguagePlugin[]): Promise<ModuleResult> {
    // 1. Collect modules from all plugins
    const modules = this.collectModules(profile, plugins);

    // 2. Enrich modules with key classes from parsed AST
    this.enrichModulesWithClasses(modules, plugins);

    // 3. Infer descriptions for modules that lack them
    this.inferDescriptions(modules);

    // 4. Extract data models from parsed files
    const dataModels = this.extractDataModels(profile, plugins);

    const result: BusinessResult = {
      modules,
      dataModels,
    };

    return result;
  }

  /**
   * Collect modules from all language plugins, deduplicating by path.
   */
  private collectModules(profile: ProjectProfile, plugins: LanguagePlugin[]): ModuleInfo[] {
    const modules: ModuleInfo[] = [];
    const seen = new Set<string>();

    for (const plugin of plugins) {
      const pluginModules = plugin.identifyModules(profile.projectPath);
      for (const mod of pluginModules) {
        const normalizedPath = path.resolve(mod.path);
        if (!seen.has(normalizedPath)) {
          seen.add(normalizedPath);
          modules.push({ ...mod, isInferred: true });
        }
      }
    }

    return modules;
  }

  /**
   * Enrich each module with key class names extracted from parsed AST nodes.
   */
  private enrichModulesWithClasses(modules: ModuleInfo[], plugins: LanguagePlugin[]): void {
    for (const mod of modules) {
      const classNames: string[] = [];

      for (const filePath of mod.keyFiles) {
        for (const plugin of plugins) {
          const nodes = plugin.parseFile(filePath);
          for (const node of nodes) {
            if (node.type === 'class' || node.type === 'interface') {
              classNames.push(node.name);
            }
          }
        }
      }

      if (classNames.length > 0) {
        mod.keyClasses = [...new Set(classNames)];
      }
    }
  }

  /**
   * Infer human-readable descriptions from directory and class names.
   */
  private inferDescriptions(modules: ModuleInfo[]): void {
    for (const mod of modules) {
      if (!mod.description || mod.description.startsWith('Module inferred from')) {
        mod.description = this.generateDescription(mod);
        mod.isInferred = true;
      }
    }
  }

  /**
   * Generate a description for a module based on its name and key classes.
   */
  private generateDescription(mod: ModuleInfo): string {
    const humanName = this.humanize(mod.name);

    if (mod.keyClasses.length > 0) {
      const classHints = mod.keyClasses.slice(0, 3).join(', ');
      return `${humanName} module (key classes: ${classHints})`;
    }

    return `${humanName} module (inferred from directory structure)`;
  }

  /**
   * Convert a directory/module name into a human-readable label.
   * e.g. "user-management" → "User Management"
   */
  private humanize(name: string): string {
    return name
      .replace(/[-_]/g, ' ')
      .split(' ')
      .filter((w) => !NOISE_WORDS.has(w.toLowerCase()))
      .map((w) => w.charAt(0).toUpperCase() + w.slice(1))
      .join(' ')
      || name;
  }

  /**
   * Extract data models (Entity/Model/DTO/VO) from all source files.
   */
  private extractDataModels(profile: ProjectProfile, plugins: LanguagePlugin[]): DataModelInfo[] {
    const models: DataModelInfo[] = [];
    const seen = new Set<string>();

    // Gather all source files from modules identified by plugins
    const allFiles = new Set<string>();
    for (const plugin of plugins) {
      const mods = plugin.identifyModules(profile.projectPath);
      for (const mod of mods) {
        for (const f of mod.keyFiles) {
          allFiles.add(f);
        }
      }
    }

    for (const filePath of allFiles) {
      for (const plugin of plugins) {
        const nodes = plugin.parseFile(filePath);
        for (const node of nodes) {
          if (node.type !== 'class' && node.type !== 'interface') continue;

          const modelType = this.classifyModelType(node.name, filePath);
          if (!modelType) continue;

          const key = `${filePath}:${node.name}`;
          if (seen.has(key)) continue;
          seen.add(key);

          const fields = this.extractFields(node);

          models.push({
            name: node.name,
            type: modelType,
            filePath,
            fields,
            description: `${modelType.toUpperCase()} class: ${node.name}`,
          });
        }
      }
    }

    return models;
  }

  /**
   * Classify a class name as entity/model/dto/vo/other based on naming patterns.
   * Returns null if the name doesn't match any data model pattern.
   */
  private classifyModelType(className: string, filePath: string): DataModelInfo['type'] | null {
    // Check class name first
    for (const { pattern, type } of DATA_MODEL_PATTERNS) {
      if (pattern.test(className)) return type;
    }

    // Check file path as fallback
    const baseName = path.basename(filePath, path.extname(filePath));
    for (const { pattern, type } of DATA_MODEL_PATTERNS) {
      if (pattern.test(baseName)) return type;
    }

    return null;
  }

  /**
   * Extract field information from AST node children.
   */
  private extractFields(node: AstNode): DataFieldInfo[] {
    const fields: DataFieldInfo[] = [];

    for (const child of node.children) {
      if (child.type !== 'field') continue;

      fields.push({
        name: child.name,
        type: child.returnType ?? 'unknown',
        annotations: child.annotations.map((a) => a.name),
        description: undefined,
      });
    }

    return fields;
  }
}
